### PR TITLE
[SPARK-25659][PYTHON][TEST] Test type inference specification for createDataFrame in PySpark

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to specify type inference and simple e2e tests. Looks we are not cleanly testing those logics. 

For instance, see https://github.com/apache/spark/blob/08c76b5d39127ae207d9d1fff99c2551e6ce2581/python/pyspark/sql/types.py#L894-L905

Looks we intended to support datetime.time and None for type inference too but it does not work:

```
>>> spark.createDataFrame([[datetime.time()]])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../spark/python/pyspark/sql/session.py", line 751, in createDataFrame
    rdd, schema = self._createFromLocal(map(prepare, data), schema)
  File "/.../spark/python/pyspark/sql/session.py", line 432, in _createFromLocal
    data = [schema.toInternal(row) for row in data]
  File "/.../spark/python/pyspark/sql/types.py", line 604, in toInternal
    for f, v, c in zip(self.fields, obj, self._needConversion))
  File "/.../spark/python/pyspark/sql/types.py", line 604, in <genexpr>
    for f, v, c in zip(self.fields, obj, self._needConversion))
  File "/.../spark/python/pyspark/sql/types.py", line 442, in toInternal
    return self.dataType.toInternal(obj)
  File "/.../spark/python/pyspark/sql/types.py", line 193, in toInternal
    else time.mktime(dt.timetuple()))
AttributeError: 'datetime.time' object has no attribute 'timetuple'
>>> spark.createDataFrame([[None]])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/.../spark/python/pyspark/sql/session.py", line 751, in createDataFrame
    rdd, schema = self._createFromLocal(map(prepare, data), schema)
  File "/.../spark/python/pyspark/sql/session.py", line 419, in _createFromLocal
    struct = self._inferSchemaFromList(data, names=schema)
  File "/.../python/pyspark/sql/session.py", line 353, in _inferSchemaFromList
    raise ValueError("Some of types cannot be determined after inferring")
ValueError: Some of types cannot be determined after inferring
```
## How was this patch tested?

Manual tests and unit tests were added.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
@BryanCutler and @viirya would you mind if I ask to take a look please?

File: python/pyspark/sql/tests.py [hunk 1]
```diff
@@ -1149,6 +1149,75 @@ def test_infer_schema(self):
...
+            1.0,
+            array.array("d", [1]),
+            [1],
+            (1, ),
+            {"a": 1},
+            bytearray(1),
+            Decimal(1),
+            Row(a=1),
+            Row("a")(1),
+            A(),
```
Comment (BryanCutler) [inline python/pyspark/sql/tests.py]: I didn't this was possible - does it just look at the variable attributes in the object to get the fields?

**Review by BryanCutler — APPROVED** [review]
LGTM, would you mind creating JIRAs for fixing inference from None and datetime.time?

File: python/pyspark/sql/tests.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/tests.py]: Yea, it uses `__dict__` attribute..  looks that's not possible in UDFs .. possibly an issue.

**Comment by HyukjinKwon** [comment]
Let me open a JIRA after taking some more further looks. I checked some code places and looks it's not easy to support `None` for instance.

**Comment by HyukjinKwon** [comment]
Merged to master.

Thank you @BryanCutler. Let me try to make sure I take further actions for related items.

**Comment by viirya** [comment]
Sorry for late. @HyukjinKwon Yes, this LGTM. It is great we can follow up other issues like None and datetime.time in other JIRA.

File: python/pyspark/sql/tests.py [hunk 2]
```diff
@@ -1149,6 +1149,75 @@ def test_infer_schema(self):
...
+
+        class A(object):
+            def __init__(self):
+                self.a = 1
+
+        data = [
+            True,
+            1,
+            "a",
+            u"a",
```
Comment (viirya) [inline python/pyspark/sql/tests.py]: nit: since this is for unicode string, how about using a non-ascii string?

Comment (HyukjinKwon) [reply python/pyspark/sql/tests.py]: Fair point. Will change when I fix some codes around here.

**Pull request closed without merging**